File: 309.best-time-to-buy-and-sell-stock-with-cooldown.cpp
/*
 * @lc app=leetcode id=309 lang=cpp
 *
 * [309] Best Time to Buy and Sell Stock with Cooldown
 */

// @lc code=start
#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        vector<vector<int>> max_sum(prices.size()+1,vector<int>(2,0));
        
        max_sum[prices.size()-1][1] = prices[prices.size()-1];
        
        for(int day = prices.size()-2; day >= 0; day--){
            for(int has = 1; has >= 0; has--){
                max_sum[day][has] = max(
                    {
                        (!has)?max_sum[day+1][!has] - prices[day]:0,
                        (has)?max_sum[day+2][!has] + prices[day]:0,
                        max_sum[day+1][has]
                    }
                );
            }
        }
        return max_sum[0][0];
    }
};
// @lc code=end

File: 309.best-time-to-buy-and-sell-stock-with-cooldown.go
/*
 * @lc app=leetcode id=309 lang=golang
 *
 * [309] Best Time to Buy and Sell Stock with Cooldown
 */

// @lc code=start
type State struct {
	day  int
	hold bool
}

var memo map[State]int

var gprices []int

func maxProfit(prices []int) int {
	gprices = prices

	memo = make(map[State]int, 0)
	return dp(0, false)
}

func dp(day int, hold bool) (mprofit int) {
	if day > len(gprices)-1 {
		return 0
	}

	state := State{day: day, hold: hold}
	if v, ok := memo[state]; ok {
		return v
	}

	mprofit = 0
	if !hold {
		mprofit = max(-gprices[day]+dp(day+1, true),
			dp(day+1, false))
	} else {
		mprofit = max(gprices[day]+dp(day+2, false),
			dp(day+1, true))
	}

	memo[state] = mprofit
	return
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

// @lc code=end

